### PR TITLE
while copying a file use file mode from the source file

### PR DESCRIPTION
**Pull request by ganesshkumar**
Signed-off-by: Ganessh Kumar R P <rpganesshkumar@gmail.com>

## Description
This is a fix to #253, where cp removes any execute permission on the file by using default `0660` on the destination file.

This commit reads the mode of the source file using `os.Stat(src)` and use the source file's permission on the destination file.  The destination file will have the same permission as the source file.

## Motivation and Context

- [x] I have raised an issue to propose this change ([#253](https://github.com/openfaas/faas-cli/issues/253))

## How Has This Been Tested?
- [x] `go build` and `go test`
- [ ]  use the build fass-cli to check the file permission (yet to do)

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- ~~My change requires a change to the documentation.~~
- ~~I have updated the documentation accordingly.~~
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- ~~I have added tests to cover my changes.~~
- [x] All ~~new and~~ existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
Hi, thanks for this change. I'm surprised we haven't seen this error yet so it's good you've caught it now.

@rgee0 can you review?

I would ideally like a unit / integration test for this too. Do we have any tests that touch this yet? cc/ @nicholasjackson 

Alex

**Comment by itscaro** [comment]
PR seems good to me

**Comment by ganesshkumar** [comment]
@itscaro  I am thinking of create a builder_test.go in the same folder and unit test the `cp` function. Do you have any other tests in mind?

**Comment by itscaro** [comment]
@ganesshkumar I think unit test for cp function is sufficient for this PR, if you can cover all cases of the `cp` function that would be great

**Comment by ganesshkumar** [comment]
@itscaro I have added tests, which tests CopyFiles function.
The test does the following,
* Creates a source folder and populates some files in it with given permission
* Creates a destination folder to copy the files to
* Uses `CopyFiles` function to copy the contents of source to destination
* Check all the destination files for the mode in which they are created

File: builder/build_test.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	fileModes := []int{0600, 0640, 0644, 0700, 0755}
+
+	dir := os.TempDir()
+	for _, mode := range fileModes {
+		// set up a source folder with 2 file
+		srcDir, srcDirErr := setupSourceFolder(2, mode)
+		if srcDirErr != nil {
+			log.Fatal("Error creating source folder")
+		}
+		defer os.RemoveAll(srcDir)
```
Comment (alexellis) [inline builder/build_test.go]: You may also want to do this up front for the case where remove didn't work or is running in parallel on the same system. Can you randomise the path any more than this?

File: builder/build_test.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"testing"
+)
+
+func Test_CopyFiles(t *testing.T) {
```
Comment (alexellis) [inline builder/build_test.go]: @itscaro thoughts on these tests?

File: builder/build_test.go (hunk 1 above)
Comment (ganesshkumar) [reply builder/build_test.go]: @alexellis, the temporary folders had some random integers attached at the end of the directory name example `openfaas-test-destination-290650684`. So I guess that will handle the case where several tests are run in parallel on the same system.

However, I notice that if a Fatal is logged then the directed till that point are not deleted. So before throwing the Fatals we should clean up all the temporary directories created. I will update the PR tomorrow with the relevant changes.

Comment (itscaro) [reply builder/build_test.go]: @ganesshkumar I prefer using t.Error() and call a function doing the teardown ops, as in fetch_template_test.go

Comment (alexellis) [reply builder/build_test.go]: @itscaro I need to merge these changes and release to unblock @ganesshkumar and a couple of other people for the Java template. Please could you raise a PR to fix the tests how you like them after this merge?

**Comment by alexellis** [comment]
@itscaro please can you raise a PR to fix up anything you think needs to be changed?

File: builder/build_test.go (hunk 2 above)
Comment (itscaro) [reply builder/build_test.go]: @ganesshkumar IMO, it's more relevant to assert file modes for ops done by `cp` function, which is in charge of doing the stuff with file modes. The test for `CopyFiles` should test only the number of files copied and the name of the destination dir.

The current test will fail on Windows, you can rename `build_test` to `build_unix_test` and add this build flag `// +build darwin dragonfly freebsd linux netbsd openbsd solaris` at the beginning of the test file

Example:
https://github.com/openfaas/faas-cli/compare/master...itscaro:pr/254?expand=1

**Comment by itscaro** [comment]
@alexellis I have raised #258 as an update for this PR

**Pull request merged**